File: src/main/java/me/soknight/minigram/chats/repository/ChatMemberRepository.java
package me.soknight.minigram.chats.repository;

import me.soknight.minigram.chats.model.entity.ChatMemberEntity;
import me.soknight.minigram.chats.model.entity.ChatMemberId;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

public interface ChatMemberRepository extends JpaRepository<ChatMemberEntity, ChatMemberId> {

    @Query("""
            select chat_member.id.profileId
            from ChatMemberEntity chat_member
            where chat_member.chat.id = :chatId
            """)
    List<UUID> findUserIdsByChatId(long chatId);

    @Query("""
            select chat_member
            from ChatMemberEntity chat_member
            where chat_member.chat.id = :chatId
            """)
    Page<ChatMemberEntity> findByChatId(long chatId, Pageable pageable);

    @Query("""
            select chat_member
            from ChatMemberEntity chat_member
            where chat_member.chat.id = :chatId and chat_member.id.profileId = :userId
            """)
    Optional<ChatMemberEntity> findById(long chatId, UUID userId);

    @Query("""
            select count(chat_member) > 0
            from ChatMemberEntity chat_member
            where chat_member.chat.id = :chatId and chat_member.id.profileId = :userId
            """)
    boolean existsById(long chatId, UUID userId);

}

File: src/main/kotlin/me/soknight/minigram/chats/repository/ChatMemberRepository.kt
@file:OptIn(ExperimentalUuidApi::class)

package me.soknight.minigram.chats.repository

import me.soknight.minigram.chats.domain.ChatMemberRow
import me.soknight.minigram.chats.model.ChatMemberRole
import me.soknight.minigram.chats.model.ChatMembersTable
import me.soknight.minigram.chats.plugin.dbQuery
import org.jetbrains.exposed.v1.core.and
import org.jetbrains.exposed.v1.core.eq
import org.jetbrains.exposed.v1.jdbc.deleteWhere
import org.jetbrains.exposed.v1.jdbc.insert
import org.jetbrains.exposed.v1.jdbc.selectAll
import kotlin.time.Clock
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid
import org.koin.core.annotation.Single

@Single
class ChatMemberRepository {

    suspend fun findUserIdsByChatId(chatId: Long): List<Uuid> = dbQuery {
        ChatMembersTable.selectAll()
            .where { ChatMembersTable.chatId eq chatId }
            .map { it[ChatMembersTable.userId] }
    }

    suspend fun findByChatId(chatId: Long, page: Int, size: Int): List<ChatMemberRow> = dbQuery {
        ChatMembersTable.selectAll()
            .where { ChatMembersTable.chatId eq chatId }
            .limit(size).offset((page * size).toLong())
            .map { it.toRow() }
    }

    suspend fun findById(chatId: Long, userId: Uuid): ChatMemberRow? = dbQuery {
        ChatMembersTable.selectAll()
            .where { (ChatMembersTable.chatId eq chatId) and (ChatMembersTable.userId eq userId) }
            .singleOrNull()?.toRow()
    }

    suspend fun existsById(chatId: Long, userId: Uuid): Boolean = dbQuery {
        ChatMembersTable.selectAll()
            .where { (ChatMembersTable.chatId eq chatId) and (ChatMembersTable.userId eq userId) }
            .count() > 0
    }

    suspend fun insert(chatId: Long, userId: Uuid, role: ChatMemberRole): ChatMemberRow = dbQuery {
        val now = Clock.System.now()
        ChatMembersTable.insert {
            it[ChatMembersTable.chatId]   = chatId
            it[ChatMembersTable.userId]   = userId
            it[ChatMembersTable.role]     = role.name
            it[ChatMembersTable.joinedAt] = now
        }
        ChatMemberRow(chatId, userId, role, now)
    }

    suspend fun delete(chatId: Long, userId: Uuid): Unit = dbQuery {
        ChatMembersTable.deleteWhere {
            (ChatMembersTable.chatId eq chatId) and (ChatMembersTable.userId eq userId)
        }
    }

    private fun org.jetbrains.exposed.v1.core.ResultRow.toRow() = ChatMemberRow(
        chatId   = this[ChatMembersTable.chatId],
        userId   = this[ChatMembersTable.userId],
        role     = ChatMemberRole.valueOf(this[ChatMembersTable.role]),
        joinedAt = this[ChatMembersTable.joinedAt]
    )
}
